### PR TITLE
Emit user notifications from chain events.

### PR DESCRIPTION
**Pull request by jnaviask**
## Description
This PR tracks the progress of the "chain notifications" project. The intent is to fetch events from various chains, process them on the server, possibly storing them in the database, and emit them as notifications to the end user.

## Motivation and Context
We already have a notification system in place, but it currently provides notifications for discussion events only. We would like to expand its scope and notify users about other important events taking place on chains they're interested in.

## How Has This Been Tested?
Unit test cases and running it with a local edgeware chain.

## TODO for MVP
- [x] Remove `manage_chain_notifications_modal` and add subscription types to the /subscriptions page.
- [x] Add proper formatting for displaying balances within notifications.
- [x] (is this possible?) Ensure that proposal event creators don't receive a notification immediately upon creation.
- [ ] ~Add webhook data.~
  - **Separate PR**.
- [ ] @raykyri Ensure interface of notification pulldown is correct -- I moved some buttons around, we should confirm their positioning.
- [ ] @raykyri Verify that url-linking when clicking on a notification is working as intended in all cases (e.g. if the proposal page no longer exists, or if we have no link available).
- [ ] @raykyri Verify subscriptions page looks OK.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] My change should be included in the release notes.
- [x] I have updated the documentation accordingly.
- [ ] I have linted the code locally prior to submission.

### CODE REVIEW

File: shared/events/edgeware/filters/labeler.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+/* eslint-disable max-len */
+const labelerFunc: LabelerFilter = (
+  blockNumber: number,
+  chainId: string,
+  data: IChainEventData,
+): IEventLabel => {
+  const balanceFormatter = (bal) => edgBalanceFormatter(chainId, bal);
+  switch (data.kind) {
+    case 'slash': {
```
Comment (drewstone) [inline shared/events/edgeware/filters/labeler.ts:37]: we should also use enums for filter cases

File: shared/events/edgeware/filters/labeler.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      const { validator, amount } = data;
+      return {
+        heading: 'Validator Slashed',
+        label: `Validator ${fmtAddr(validator)} was slashed by amount ${balanceFormatter(amount)}.`,
+      };
+    }
+    case 'reward': {
```
Comment (drewstone) [inline shared/events/edgeware/filters/labeler.ts:44]: here too and so on

File: shared/events/edgeware/filters/titler.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import { SubstrateEventKind } from '../types';
+import { IEventTitle, TitlerFilter } from '../../interfaces';
+
+const titlerFunc: TitlerFilter = (kind: SubstrateEventKind): IEventTitle => {
+  switch (kind) {
+    case 'slash': {
```
Comment (drewstone) [inline shared/events/edgeware/filters/titler.ts:6]: here too and so on

File: shared/events/edgeware/filters/type_parser.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { Event } from '@polkadot/types/interfaces';
+import { SubstrateEventKind } from '../types';
+
+export default function (event: Event): SubstrateEventKind | null {
+  switch (event.section) {
+    case 'staking':
```
Comment (drewstone) [inline shared/events/edgeware/filters/type_parser.ts:6]: here too and so on

File: shared/events/edgeware/types.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * To implement a new form of event, add it to this enum, and add its
+ * JSON interface below (ensure it is stringify-able and then parse-able).
+ */
+
+/** Special types for formatting/labeling purposes */
+export type SubstrateBalanceString = string;
+export type SubstrateBlockNumber = number;
+export type SubstrateAccountId = string;
+
+export interface ISubstrateSlash {
```
Comment (drewstone) [inline shared/events/edgeware/types.ts:22]: seems we can use these as the enums? Is it possible even though they're interfaces?

```
switch (thing) {
  case ISubstrateSlash.kind:
    ...
}
```

File: client/scripts/views/components/header.ts [hunk 6]
```diff
@@ -681,6 +686,35 @@ const HeaderNotificationRow: m.Component<IHeaderNotificationRow> = {
           + `${slugify(decoded_title)}`,
         () => jumpHighlightComment('parent')
       );
+    } else if (category === NotificationCategories.ChainEvent) {
+      // TODO: this needs to be improved a lot lol
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply client/scripts/views/components/header.ts:690]: comment isn't relevant anymore, what's beneath it is the significantly improved version, but i forgot to remove the TODO

File: client/scripts/views/pages/subscriptions.ts [hunk 7]
```diff
@@ -308,6 +312,86 @@ const CommunitySubscriptions: m.Component<{}, ICommunitySubscriptionsState> = {
...
+
+interface IEventSubscriptionState {
+  chain: string;
+  eventKinds: IChainEventKind[];
+  allSupportedChains: string[];
+}
+
+const EventSubscriptions: m.Component<{}, IEventSubscriptionState> = {
+  oninit: (vnode) => {
+    vnode.state.chain = 'edgeware';
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply client/scripts/views/pages/subscriptions.ts:358]: that's what the dropdown is for

[the remark this answers is not in the document]
Comment (jnaviask) [reply client/scripts/views/pages/subscriptions.ts:358]: in theory, the conditional will select the correct list of event kinds, which then populate the rows

File: eventSubscriber.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import subscribeEdgewareEvents from './shared/events/edgeware/index';
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply eventSubscriber.ts:1]: yeah, it's a demo basically, could be useful if we're thinking of moving these changes to another repo, but we can move it elsewhere for now (although not sure where)

File: server/scripts/resetServer.ts [hunk 9]
```diff
@@ -3,14 +3,15 @@ import { NotificationCategories } from '../../shared/types';
...
+import { SubstrateEventKinds } from '../../shared/events/edgeware/types';
 
 const nodes = [
   [ 'localhost:9944', 'edgeware-local' ],
   [ 'berlin1.edgewa.re', 'edgeware-testnet' ],
   [ 'berlin2.edgewa.re', 'edgeware-testnet' ],
   [ 'berlin3.edgewa.re', 'edgeware-testnet' ],
-  [ 'mainnet1.edgewa.re', 'edgeware' ],
-  //[ 'localhost:9944', 'kusama-local' ],
+  [ 'mainnet1.edgewa.re:9944', 'edgeware' ],
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply server/scripts/resetServer.ts:13]: i recall it breaking for me without the port, but i can try removing it and see

File: shared/events/edgeware/filters/enricher.ts [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+  Event, ReferendumInfoTo239, AccountId, TreasuryProposal, Balance, PropIndex,
+  ReferendumIndex, ProposalIndex
+} from '@polkadot/types/interfaces';
+import { Option, bool } from '@polkadot/types';
+import { Codec } from '@polkadot/types/types';
+import { SubstrateEventKind } from '../types';
+import { CWEvent } from '../../interfaces';
+
+// TODO: better balance/BN handling than string
+export default async function (
```
[the remark this answers is not in the document]
Comment (jnaviask) [reply shared/events/edgeware/filters/enricher.ts:12]: will do -- should also remove that TODO

File: shared/events/edgeware/types.ts (hunk 5 above)
Comment (jnaviask) [reply shared/events/edgeware/types.ts:22]: I don't think so, you can't use a type (interfaces are types) as a value. But I could swap out all the "kind" strings with a string enum.

File: server/scripts/resetServer.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jnaviask) [reply server/scripts/resetServer.ts:13]: Yes -- it needs the port, unfortunately.

File: eventSubscriber.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jnaviask) [reply eventSubscriber.ts:1]: where do you think we should move it?

**Comment by drewstone** [comment]
Can we list here in this doc what events are working? Would help run through QA process.

**Pull request merged**